File: src/stats.ts
import type { Config } from "./config.ts";
import { parseGoalDate } from "./config.ts";
import type { Workout } from "./models.ts";
import { calendarDay, formatSeconds, pace500mSeconds, parsedDate } from "./models.ts";

export interface WeekSummary {
  weekStart: Date;
  meters: number;
  sessions: number;
  paceSum: number;
  paceCount: number;
  spmSum: number;
  spmCount: number;
  hrSum: number;
  hrCount: number;
}

export interface GoalProgress {
  target: number;
  totalMeters: number;
  progress: number;
  weeksElapsed: number;
  totalWeeks: number;
  remainingMeters: number;
  remainingWeeks: number;
  requiredPace: number;
  currentAvgPace: number;
  onPace: boolean;
}

const RECENT_WEEKS = 4;

export function mondayOf(t: Date): Date {
  const d = new Date(t.getFullYear(), t.getMonth(), t.getDate());
  const offset = (d.getDay() + 6) % 7;
  d.setDate(d.getDate() - offset);
  return d;
}

export function workoutsInRange(workouts: Workout[], from: Date, to: Date): Workout[] {
  return workouts.filter((w) => {
    const t = parsedDate(w);
    return t >= from && t < to;
  });
}

export function buildWeekSummaries(workouts: Workout[], now: Date, weeks: number): WeekSummary[] {
  const thisMonday = mondayOf(now);
  const cutoff = new Date(thisMonday);
  cutoff.setDate(cutoff.getDate() - (weeks - 1) * 7);

  const summaries: WeekSummary[] = [];
  for (let i = 0; i < weeks; i++) {
    const ws = new Date(thisMonday);
    ws.setDate(ws.getDate() - (weeks - 1 - i) * 7);
    summaries.push({
      weekStart: ws,
      meters: 0,
      sessions: 0,
      paceSum: 0,
      paceCount: 0,
      spmSum: 0,
      spmCount: 0,
      hrSum: 0,
      hrCount: 0,
    });
  }

  const daysByWeek = new Map<number, Set<string>>();

  for (const w of workouts) {
    const t = new Date(w.date.replace(" ", "T"));
    if (t < cutoff || t > now) continue;

    const monday = mondayOf(t);
    const diffDays = Math.round((monday.getTime() - cutoff.getTime()) / (1000 * 60 * 60 * 24));
    const idx = Math.floor(diffDays / 7);
    if (idx < 0 || idx >= weeks) continue;

    const ws = summaries[idx]!;
    ws.meters += w.distance;

    if (!daysByWeek.has(idx)) daysByWeek.set(idx, new Set());
    daysByWeek.get(idx)!.add(calendarDay(w));

    const pace = pace500mSeconds(w);
    if (pace > 0) {
      ws.paceSum += pace;
      ws.paceCount++;
    }
    if (w.stroke_rate && w.stroke_rate > 0) {
      ws.spmSum += w.stroke_rate;
      ws.spmCount++;
    }
    if (w.heart_rate?.average && w.heart_rate.average > 0) {
      ws.hrSum += w.heart_rate.average;
      ws.hrCount++;
    }
  }

  for (const [idx, days] of daysByWeek) {
    summaries[idx]!.sessions = days.size;
  }

  return summaries;
}

export interface RecentWeek {
  weekStart: Date;
  meters: number;
  sessions: number;
}

export function recentWeeks(workouts: Workout[], now: Date, count: number): RecentWeek[] {
  const out: RecentWeek[] = [];
  for (let i = 0; i < count; i++) {
    const weekStart = mondayOf(now);
    weekStart.setDate(weekStart.getDate() - i * 7);
    const weekEnd = new Date(weekStart);
    weekEnd.setDate(weekEnd.getDate() + 7);
    const weekWorkouts = workoutsInRange(workouts, weekStart, weekEnd);
    out.push({
      weekStart,
      meters: weekWorkouts.reduce((sum, w) => sum + w.distance, 0),
      sessions: new Set(weekWorkouts.map(calendarDay)).size,
    });
  }
  return out;
}

export interface WeekSummaryData {
  week_start: string;
  meters: number;
  sessions: number;
  avg_pace_500m_seconds: number | null;
  avg_pace_500m: string | null;
  avg_spm: number | null;
  avg_hr: number | null;
}

export function localYMD(d: Date): string {
  const mm = String(d.getMonth() + 1).padStart(2, "0");
  const dd = String(d.getDate()).padStart(2, "0");
  return `${d.getFullYear()}-${mm}-${dd}`;
}

export function weekSummaryData(ws: WeekSummary): WeekSummaryData {
  const avgPace = ws.paceCount > 0 ? ws.paceSum / ws.paceCount : null;
  return {
    week_start: localYMD(ws.weekStart),
    meters: ws.meters,
    sessions: ws.sessions,
    avg_pace_500m_seconds: avgPace != null ? Math.round(avgPace * 10) / 10 : null,
    avg_pace_500m: avgPace != null ? formatSeconds(avgPace) : null,
    avg_spm: ws.spmCount > 0 ? Math.round((ws.spmSum / ws.spmCount) * 10) / 10 : null,
    avg_hr: ws.hrCount > 0 ? Math.round(ws.hrSum / ws.hrCount) : null,
  };
}

export function computeGoalProgress(workouts: Workout[], cfg: Config, now?: Date): GoalProgress {
  const target = cfg.goal.target_meters;
  const start = parseGoalDate(cfg.goal.start_date);
  const end = parseGoalDate(cfg.goal.end_date);
  const today = now ?? new Date();

  let totalMeters = 0;
  for (const w of workouts) {
    const t = parsedDate(w);
    if (t >= start && t <= end) {
      totalMeters += w.distance;
    }
  }

  const progress = totalMeters / target;
  const totalDays = (end.getTime() - start.getTime()) / (1000 * 60 * 60 * 24);
  const totalWeeks = Math.ceil(totalDays / 7);

  let weeksElapsed = 0;
  if (today > start) {
    weeksElapsed = Math.floor((today.getTime() - start.getTime()) / (1000 * 60 * 60 * 24 * 7));
  }

  let remainingMeters = target - totalMeters;
  if (remainingMeters < 0) remainingMeters = 0;
  let remainingWeeks = totalWeeks - weeksElapsed;
  if (remainingWeeks < 1) remainingWeeks = 1;
  const requiredPace = Math.floor(remainingMeters / remainingWeeks);

  let currentAvgPace = 0;
  if (weeksElapsed > 0) {
    const thisMonday = mondayOf(today);
    const recentStart = new Date(thisMonday);
    recentStart.setDate(recentStart.getDate() - RECENT_WEEKS * 7);
    const windowStart = recentStart < start ? start : recentStart;
    const windowMs = thisMonday.getTime() - windowStart.getTime();
    const weeksInWindow = Math.max(1, Math.round(windowMs / (1000 * 60 * 60 * 24 * 7)));
    let recentMeters = 0;
    for (const w of workouts) {
      const t = parsedDate(w);
      if (t >= windowStart && t < thisMonday) {
        recentMeters += w.distance;
      }
    }
    currentAvgPace = Math.floor(recentMeters / weeksInWindow);
  }
  const targetWeekly = target / totalWeeks;
  const onPace = currentAvgPace >= targetWeekly;

  return {
    target,
    totalMeters,
    progress,
    weeksElapsed,
    totalWeeks,
    remainingMeters,
    remainingWeeks,
    requiredPace,
    currentAvgPace,
    onPace,
  };
}

File: internal/stats/stats.go
package stats

import (
	"fmt"
	"math"
	"time"

	"github.com/richhaase/c2/internal/config"
	"github.com/richhaase/c2/internal/models"
)

type WeekSummary struct {
	WeekStart time.Time
	Meters    int
	Sessions  int
	PaceSum   float64
	PaceCount int
	SPMSum    int
	SPMCount  int
	HRSum     int
	HRCount   int
}

type GoalProgress struct {
	Target          int     `json:"target"`
	TotalMeters     int     `json:"totalMeters"`
	Progress        float64 `json:"progress"`
	WeeksElapsed    int     `json:"weeksElapsed"`
	TotalWeeks      int     `json:"totalWeeks"`
	RemainingMeters int     `json:"remainingMeters"`
	RemainingWeeks  int     `json:"remainingWeeks"`
	RequiredPace    int     `json:"requiredPace"`
	CurrentAvgPace  int     `json:"currentAvgPace"`
	OnPace          bool    `json:"onPace"`
}

type RecentWeek struct {
	WeekStart time.Time
	Meters    int
	Sessions  int
}

type WeekSummaryData struct {
	WeekStart          string   `json:"week_start"`
	Meters             int      `json:"meters"`
	Sessions           int      `json:"sessions"`
	AvgPace500mSeconds *float64 `json:"avg_pace_500m_seconds"`
	AvgPace500m        *string  `json:"avg_pace_500m"`
	AvgSPM             *float64 `json:"avg_spm"`
	AvgHR              *int     `json:"avg_hr"`
}

type GoalProjection struct {
	RemainingWeeks       float64 `json:"remaining_weeks"`
	ProjectedTotalMeters int     `json:"projected_total_meters"`
	ProjectedPct         float64 `json:"projected_pct"`
	ShortfallMeters      int     `json:"shortfall_meters"`
}

const (
	recentPaceWeeks = 4
	daysPerWeek     = 7
	hoursPerWeek    = 24 * daysPerWeek
	secondsPerDay   = 24 * 60 * 60
)

func SessionCount(workouts []models.Workout) int {
	days := make(map[string]struct{}, len(workouts))
	for _, w := range workouts {
		days[models.CalendarDay(w)] = struct{}{}
	}
	return len(days)
}

func MondayOf(t time.Time) time.Time {
	y, m, d := t.In(time.Local).Date()
	midnight := time.Date(y, m, d, 0, 0, 0, 0, time.Local)
	offset := (int(midnight.Weekday()) + 6) % daysPerWeek
	return midnight.AddDate(0, 0, -offset)
}

func WorkoutsInRange(workouts []models.Workout, from, to time.Time) []models.Workout {
	out := make([]models.Workout, 0, len(workouts))
	for _, w := range workouts {
		t := models.ParsedDate(w)
		if !t.Before(from) && t.Before(to) {
			out = append(out, w)
		}
	}
	return out
}

func BuildWeekSummaries(workouts []models.Workout, now time.Time, weeks int) []WeekSummary {
	thisMonday := MondayOf(now)
	cutoff := thisMonday.AddDate(0, 0, -(weeks-1)*daysPerWeek)

	summaries := make([]WeekSummary, 0, weeks)
	for i := 0; i < weeks; i++ {
		summaries = append(summaries, WeekSummary{
			WeekStart: thisMonday.AddDate(0, 0, -(weeks-1-i)*daysPerWeek),
		})
	}

	daysByWeek := make(map[int]map[string]struct{})

	for _, w := range workouts {
		t := models.ParsedDate(w)
		if t.Before(cutoff) || t.After(now) {
			continue
		}

		idx := floorDiv(dayNumber(MondayOf(t))-dayNumber(cutoff), daysPerWeek)
		if idx < 0 || idx >= weeks {
			continue
		}

		ws := &summaries[idx]
		ws.Meters += w.Distance

		days, ok := daysByWeek[idx]
		if !ok {
			days = make(map[string]struct{})
			daysByWeek[idx] = days
		}
		days[models.CalendarDay(w)] = struct{}{}

		if pace := models.Pace500mSeconds(w); pace > 0 {
			ws.PaceSum += pace
			ws.PaceCount++
		}
		if w.StrokeRate != nil && *w.StrokeRate > 0 {
			ws.SPMSum += *w.StrokeRate
			ws.SPMCount++
		}
		if w.HeartRate != nil && w.HeartRate.Average != nil && *w.HeartRate.Average > 0 {
			ws.HRSum += *w.HeartRate.Average
			ws.HRCount++
		}
	}

	for idx, days := range daysByWeek {
		summaries[idx].Sessions = len(days)
	}

	return summaries
}

func RecentWeeks(workouts []models.Workout, now time.Time, count int) []RecentWeek {
	out := make([]RecentWeek, 0, count)
	for i := 0; i < count; i++ {
		weekStart := MondayOf(now).AddDate(0, 0, -i*daysPerWeek)
		weekEnd := weekStart.AddDate(0, 0, daysPerWeek)
		weekWorkouts := WorkoutsInRange(workouts, weekStart, weekEnd)
		meters := 0
		for _, w := range weekWorkouts {
			meters += w.Distance
		}
		out = append(out, RecentWeek{
			WeekStart: weekStart,
			Meters:    meters,
			Sessions:  SessionCount(weekWorkouts),
		})
	}
	return out
}

func LocalYMD(t time.Time) string {
	y, m, d := t.In(time.Local).Date()
	return fmt.Sprintf("%04d-%02d-%02d", y, int(m), d)
}

func WeekSummaryDataOf(ws WeekSummary) WeekSummaryData {
	out := WeekSummaryData{
		WeekStart: LocalYMD(ws.WeekStart),
		Meters:    ws.Meters,
		Sessions:  ws.Sessions,
	}
	if ws.PaceCount > 0 {
		avgPace := ws.PaceSum / float64(ws.PaceCount)
		seconds := math.Round(avgPace*10) / 10
		formatted := models.FormatSeconds(avgPace)
		out.AvgPace500mSeconds = &seconds
		out.AvgPace500m = &formatted
	}
	if ws.SPMCount > 0 {
		spm := math.Round(float64(ws.SPMSum)/float64(ws.SPMCount)*10) / 10
		out.AvgSPM = &spm
	}
	if ws.HRCount > 0 {
		hr := int(math.Round(float64(ws.HRSum) / float64(ws.HRCount)))
		out.AvgHR = &hr
	}
	return out
}

func ProjectGoal(goal GoalProgress, end, now time.Time) GoalProjection {
	weeksLeft := calendarDaysBetween(now, end) / daysPerWeek
	return projectGoal(goal, weeksLeft)
}

func ProjectGoalByElapsedTime(goal GoalProgress, end, now time.Time) GoalProjection {
	weeksLeft := end.Sub(now).Hours() / hoursPerWeek
	return projectGoal(goal, weeksLeft)
}

func projectGoal(goal GoalProgress, weeksLeft float64) GoalProjection {
	if weeksLeft < 0 {
		weeksLeft = 0
	}
	projected := int(math.Round(float64(goal.TotalMeters) + float64(goal.CurrentAvgPace)*weeksLeft))
	shortfall := goal.Target - projected
	if shortfall < 0 {
		shortfall = 0
	}
	projectedPct := 0.0
	if goal.Target > 0 {
		projectedPct = math.Round(float64(projected)/float64(goal.Target)*1000) / 10
	}
	return GoalProjection{
		RemainingWeeks:       math.Round(weeksLeft*10) / 10,
		ProjectedTotalMeters: projected,
		ProjectedPct:         projectedPct,
		ShortfallMeters:      shortfall,
	}
}

func ComputeGoalProgress(workouts []models.Workout, cfg config.Config, now time.Time) (GoalProgress, error) {
	target := cfg.Goal.TargetMeters
	if target <= 0 {
		return GoalProgress{}, fmt.Errorf("Goal target must be a positive number of meters.")
	}
	start, err := config.ParseGoalDate(cfg.Goal.StartDate)
	if err != nil {
		return GoalProgress{}, err
	}
	end, err := config.ParseGoalDate(cfg.Goal.EndDate)
	if err != nil {
		return GoalProgress{}, err
	}
	if end.Before(start) {
		return GoalProgress{}, fmt.Errorf("Goal end date must not be before start date.")
	}
	endExclusive := end.AddDate(0, 0, 1)
	today := now
	if today.IsZero() {
		today = time.Now()
	}

	totalMeters := 0
	for _, w := range workouts {
		t := models.ParsedDate(w)
		if !t.Before(start) && t.Before(endExclusive) {
			totalMeters += w.Distance
		}
	}

	progress := float64(totalMeters) / float64(target)
	totalDays := dayNumber(endExclusive) - dayNumber(start)
	totalWeeks := int(math.Ceil(float64(totalDays) / daysPerWeek))

	weeksElapsed := 0
	if today.After(start) {
		weeksElapsed = floorDiv(dayNumber(today)-dayNumber(start), daysPerWeek)
		if weeksElapsed > totalWeeks {
			weeksElapsed = totalWeeks
		}
	}

	remainingMeters := target - totalMeters
	if remainingMeters < 0 {
		remainingMeters = 0
	}
	remainingWeeks := totalWeeks - weeksElapsed
	if remainingWeeks < 1 {
		remainingWeeks = 1
	}
	requiredPace := int(math.Floor(float64(remainingMeters) / float64(remainingWeeks)))

	currentAvgPace := 0
	if weeksElapsed > 0 {
		thisMonday := MondayOf(today)
		windowStart := thisMonday.AddDate(0, 0, -recentPaceWeeks*daysPerWeek)
		if windowStart.Before(start) {
			windowStart = start
		}
		weeksInWindow := math.Round(float64(dayNumber(thisMonday)-dayNumber(windowStart)) / daysPerWeek)
		if weeksInWindow < 1 {
			weeksInWindow = 1
		}
		recentMeters := 0
		for _, w := range workouts {
			t := models.ParsedDate(w)
			if !t.Before(windowStart) && t.Before(thisMonday) {
				recentMeters += w.Distance
			}
		}
		currentAvgPace = int(math.Floor(float64(recentMeters) / weeksInWindow))
	}
	targetWeekly := float64(target) / float64(totalWeeks)

	return GoalProgress{
		Target:          target,
		TotalMeters:     totalMeters,
		Progress:        progress,
		WeeksElapsed:    weeksElapsed,
		TotalWeeks:      totalWeeks,
		RemainingMeters: remainingMeters,
		RemainingWeeks:  remainingWeeks,
		RequiredPace:    requiredPace,
		CurrentAvgPace:  currentAvgPace,
		OnPace:          float64(currentAvgPace) >= targetWeekly,
	}, nil
}

func dayNumber(t time.Time) int {
	y, m, d := t.In(time.Local).Date()
	return int(time.Date(y, m, d, 0, 0, 0, 0, time.UTC).Unix() / secondsPerDay)
}

func floorDiv(a, b int) int {
	q := a / b
	if a%b != 0 && (a < 0) != (b < 0) {
		q--
	}
	return q
}

func calendarDaysBetween(from, to time.Time) float64 {
	return float64(dayNumber(to)-dayNumber(from)) + dayFraction(to) - dayFraction(from)
}

func dayFraction(t time.Time) float64 {
	local := t.In(time.Local)
	seconds := local.Hour()*60*60 + local.Minute()*60 + local.Second()
	return (float64(seconds) + float64(local.Nanosecond())/float64(time.Second)) / secondsPerDay
}
